speaker/bgp_controller.go

```diff
@@ -272,20 +447,301 @@ type session interface {
 func (c *bgpController) SetLeader(log.Logger, bool) {}
 
 func (c *bgpController) SetNode(l log.Logger, node *v1.Node) error {
+	nodeAnnotations := node.Annotations
+	if nodeAnnotations == nil {
+		nodeAnnotations = map[string]string{}
+	}
 	nodeLabels := node.Labels
 	if nodeLabels == nil {
 		nodeLabels = map[string]string{}
 	}
-	ns := labels.Set(nodeLabels)
-	if c.nodeLabels != nil && labels.Equals(c.nodeLabels, ns) {
-		// Node labels unchanged, no action required.
+
+	anns := labels.Set(nodeAnnotations)
+	ls := labels.Set(nodeLabels)
+	annotationsUnchanged := c.nodeAnnotations != nil && labels.Equals(c.nodeAnnotations, anns)
+	labelsUnchanged := c.nodeLabels != nil && labels.Equals(c.nodeLabels, ls)
+	if labelsUnchanged && annotationsUnchanged {
+		// Node labels and annotations unchanged, no action required.
 		return nil
 	}
-	c.nodeLabels = ns
-	l.Log("event", "nodeLabelsChanged", "msg", "Node labels changed, resyncing BGP peers")
+	c.nodeAnnotations = anns
+	c.nodeLabels = ls
+
+	c.discoverNodePeer(l)
+
+	l.Log("event", "nodeChanged", "msg", "node changed, resyncing BGP peers")
 	return c.syncPeers(l)
 }
 
-var newBGP = func(logger log.Logger, addr string, myASN uint32, routerID net.IP, asn uint32, hold time.Duration, password string, myNode string) (session, error) {
-	return bgp.New(logger, addr, myASN, routerID, asn, hold, password, myNode)
+func (c *bgpController) StatusHandler() func(w http.ResponseWriter, r *http.Request) {
```

Comment:
I assume you've considered the thread-safety of this method and determined no locking is needed?

Reply:
Huh. I think there might be a race between the endpoint and the control loop :scream:

Up to now no locking was needed in the speaker since AFAICT we process k8s events serially:

https://github.com/metallb/metallb/blob/25a7d8d1b6bb96695035265f7e60428f3c4f48df/internal/k8s/k8s.go#L304-L321

**However**, what happens if someone queries the status endpoint during e.g. a config change? I'm looking into it.

Reply:
I can't seem to get it to crash by querying the endpoint in a loop and making config changes in parallel. I'll look for a better way to test this. Maybe we can use the Go race detector.

Reply:
It seems reasonable if it doesn't crash, though (is not a call to `free()` and you are reading that after). As you mentioned, go race detector might be better.

I'll review this later, but my guess is that some synchronization is need

Reply:
Yeah, sorry for not providing better comments on this.  I mostly wanted to flag it to make sure we all agreed it was safe.  I wanted to come back and look closer, too

Reply:
The peers slice is protected by a mutex now.